README.md
ntpd\_driver
============

This ROS node listen `sensor_msgs/TimeReference` and send it to ntpd via SHM (like gpsd).

Parameter `~/shm_unit` define SHM unit (in ntp.conf) (int, default: 2).
Parameter `~/fixup_date` enable/disable date fixup (bool, default: false)


System configuration
--------------------

### ntpd configuration

Add this to `/etc/ntp.conf`:

    ### GPS SHM driver
    server 127.127.28.2 minpoll 4 maxpoll 4
    fudge 127.127.28.2 time1 0.5 stratum 12 refid ROS

And then restart ntp service.

Run example:

    rosrun ntpd_driver shm_driver _shm_unit:=2 _time_ref_topic:=/mavros/time_reference


### chrony configuration

Add this to `/etc/chrony/chrony.conf`:

    ### SHM driver
    refclock SHM 0 delay 0.5 refid ROS

And then restart chrony service.

Run example:

    rosrun ntpd_driver shm_driver _shm_unit:=0 _time_ref_topic:=/mavros/time_reference


### Date fixup configuration (sudo)

On my Raspberry Pi 2 ntpd reject SHM data if system date is not set (e.g. JAN 1970).
To fix that `shm_driver` now can set system time if it unset.

For setting date program requires root privileges, so used `sudo`.

Add this to `/etc/sudoers` (using `visudo`):

    %sudo	ALL=NOPASSWD: /bin/date



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package ntpd_driver
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

2.2.0 (2022-12-30)
------------------
* include: fix include warnings, arain
* fix ament_cpplint warnings
* node: remove time_ref_topic parameter, port for for `#9 <https://github.com/vooon/ntpd_driver/issues/9>`_
* Merge branch 'master' into ros2
  * master:
  1.3.0
  update changelog
  ci: return semgrep default rule set
  ci: update ros-i ci action
  ci: add industrial CI
  shm_driver: remove time_ref topic parameter to improve secutiry, please use remap
  add license file
* 1.3.0
* update changelog
* ci: return semgrep default rule set
* ci: update ros-i ci action
* ci: add industrial CI
* shm_driver: remove time_ref topic parameter to improve secutiry, please use remap
  Fixes `#9 <https://github.com/vooon/ntpd_driver/issues/9>`_
* add license file
* Contributors: Vladimir Ermakov

2.1.0 (2021-03-25)
------------------
* make linker happy
* fix loading as a component
* Contributors: Vladimir Ermakov

2.0.2 (2021-02-26)
------------------
* make tests happy
* license: change formatting to one known by ament_copyright
* ci: initial import
* add license file
* Contributors: Vladimir Ermakov

2.0.1 (2019-12-29)
------------------
* disable depend on launch_xml, rosdep can't find it
* Contributors: Vladimir Ermakov

2.0.0 (2019-12-29)
------------------
* update launch
* read parameters
* subscribe to time_ref topic, remove log THROTTLE - segfaults
* open shm
* make it buildable, todo: subscribe, open shm
* update cmake to use ament
* Contributors: Vladimir Ermakov

1.3.0 (2022-12-30)
------------------
* ci: return semgrep default rule set
* ci: update ros-i ci action
* ci: add industrial CI
* shm_driver: remove time_ref topic parameter to improve secutiry, please use remap
  Fixes `#9 <https://github.com/vooon/ntpd_driver/issues/9>`_
* add license file
* Contributors: Vladimir Ermakov

1.2.0 (2015-10-02)
------------------
* `#2 <https://github.com/vooon/ntpd_driver/issues/2>`_: allow both UDPROS and TCPROS
* Contributors: Vladimir Ermakov

1.1.1 (2015-07-18)
------------------
* Syncronize shm write with gpsd/ntpshmwrite.c
* Add fixup_date parameter.
  Now driver can setup system date.
* Contributors: Vladimir Ermakov

1.1.0 (2015-04-17)
------------------
* Update package.xml for REP 140
* Change topic param name to `~/time_ref_topic`
* Merge pull request `#1 <https://github.com/vooon/ntpd_driver/issues/1>`_ from oceansystemslab/master
  Updated Driver
* Updated README with new node input parameter
* fixed indentation according to ROS guidelines and added input parameter for time reference topic
* Removed transport hints to allow the driver to work with the nmea_serial_driver Python publisher and added ROS style indentation.
* Updated .gitignore with C++ version on GitHub
* Contributors: Valerio De Carolis, Vladimir Ermakov

1.0.2 (2015-01-24)
------------------
* Add example launch script.
* Add rosindex metadata.
* Contributors: Vladimir Ermakov

1.0.1 (2014-07-13)
------------------
* Rename package from ros2ntpd to ntpd_driver
  Add notes for chrony.
* Contributors: Vladimir Ermakov

1.0.0 (2014-06-28)
------------------
* Update readme
* Add readme
* Update shmTime struct to latest version.
  Fix: change leap=0 will fix ntpd receive.
* Initial
* Contributors: Vladimir Ermakov


src/ros2ntpd.cpp
/** ROS ntpd driver (via shm)
 *
 * This node will subscribe to any TimeReference message
 * and sends it's data to ntpd.
 *
 * @author Vladimir Ermakov <vooon341@gmail.com>
 * @license BSD
 *
 * Copyright 2014 Vladimir Ermakov.
 * Based on ntpshm.c from gpsd. 
 */


#include <ros/ros.h>
#include <ros/console.h>

#include <sensor_msgs/TimeReference.h>

#include <sys/ipc.h>
#include <sys/shm.h>

#include <signal.h>
#include <cstring>
#include <cerrno>

/** the definition of shmTime is from ntpd source ntpd/refclock_shm.c */
struct shmTime
{
	int mode;	/* 0 - if valid set
			 *       use values,
			 *       clear valid
			 * 1 - if valid set
			 *       if count before and after read of values is equal,
			 *         use values
			 *       clear valid
			 */
	volatile int count;
	time_t clockTimeStampSec;
	int clockTimeStampUSec;
	time_t receiveTimeStampSec;
	int receiveTimeStampUSec;
	int leap;
	int precision;
	int nsamples;
	volatile int valid;
	unsigned        clockTimeStampNSec;     /* Unsigned ns timestamps */
	unsigned        receiveTimeStampNSec;   /* Unsigned ns timestamps */
	int             dummy[8];
};

const long int NTPD_SHM_BASE = 0x4e545030;

/** Map SHM page
 *
 * NOTE: this function did not create SHM like gpsd/ntpshm.c
 */
static volatile struct shmTime *get_shmTime(int unit)
{
	/* we definitly not root, no IPC_CREAT */
	const int get_flags = 0666;

	int shmid = shmget(NTPD_SHM_BASE + unit, sizeof(shmTime), get_flags);
	if (shmid < 0) {
		ROS_FATAL("SHM shmget(0x%08lx, %zd, %o) fail: %s",
				NTPD_SHM_BASE + unit,
				sizeof(shmTime),
				get_flags,
				strerror(errno));
		return NULL;
	}

	/* TODO: check noumber of attached progs */

	void *p = shmat(shmid, 0, 0);
	if (p == (void *) -1) {
		ROS_FATAL("SHM shmat(%d, 0, 0) fail: %s", shmid, strerror(errno));
		return NULL;
	}

	ROS_INFO("SHM(%d) key 0x%08lx, successfully opened",
			unit, NTPD_SHM_BASE + unit);
	return (volatile struct shmTime*) p;
}

static void put_shmTime(volatile struct shmTime **shm)
{
	ROS_ASSERT(shm != NULL);
	if (*shm == NULL)
		return;

	if (shmdt((const void*) *shm) == -1)
		ROS_FATAL("SHM shmdt(%p) fail: %s", *shm, strerror(errno));
	else
		*shm = NULL;
}


/** global SHM time handle */
volatile struct shmTime *g_shm = NULL;

static void sig_handler(int sig)
{
	put_shmTime(&g_shm);
	ros::shutdown();
}

static void time_ref_cb(const sensor_msgs::TimeReference::ConstPtr &time_ref)
{
	if (g_shm == NULL) {
		ROS_FATAL("Got time_ref before shm opens.");
		ros::shutdown();
	}

	/* header */
	g_shm->mode = 1;
	g_shm->leap = 0;	// LEAP_NOWARNING
	g_shm->precision = -1;	// initially 0.5 sec
	g_shm->nsamples = 3;	// stages of median filter

	/* ntpshm.c recommends add barrier before inc count */
	g_shm->valid = 0;
	g_shm->count += 1;
	g_shm->receiveTimeStampSec = time_ref->header.stamp.sec;
	g_shm->receiveTimeStampUSec = time_ref->header.stamp.nsec / 1000;
	g_shm->receiveTimeStampNSec = time_ref->header.stamp.nsec;
	g_shm->clockTimeStampSec = time_ref->time_ref.sec;
	g_shm->clockTimeStampUSec = time_ref->time_ref.nsec / 1000;
	g_shm->clockTimeStampNSec = time_ref->time_ref.nsec;
	/* barrier again */
	g_shm->count += 1;
	g_shm->valid = 1;

	ROS_DEBUG_THROTTLE(10, "Got time_ref: %lu.%09lu",
			(long unsigned) time_ref->time_ref.sec,
			(long unsigned) time_ref->time_ref.nsec);
}

int main(int argc, char *argv[])
{
	ros::init(argc, argv, "ros2ntpd");
	ros::NodeHandle nh("~");
	ros::Subscriber time_ref_sub;
	int shm_unit;

	// Override default ROS handler
	signal(SIGINT, sig_handler);

	nh.param("shm_unit", shm_unit, 2);
	g_shm = get_shmTime(shm_unit);
	if (g_shm == NULL)
		return 1;

	time_ref_sub = nh.subscribe("time_ref", 10, time_ref_cb,
			ros::TransportHints()
				.unreliable()
				.maxDatagramSize(1024));

	ros::spin();
	put_shmTime(&g_shm);
	return 0;
}


src/ShmDriver.cpp

#include <ntpd_driver/ShmDriver.hpp>
#include <class_loader/register_macro.hpp>

#include <sys/ipc.h>
#include <sys/shm.h>

#include <cstring>

// there may be another library, but Poco already used by class_loader,
// so it definetly exist in your system.
#include <Poco/Process.h>
#include <Poco/Pipe.h>
#include <Poco/PipeStream.h>
#include <Poco/Format.h>
#include <Poco/StreamCopier.h>

/**
 * Memory barrier. unfortunatly we can't use C stdatomic.h
 * So only one option: asm magick
 *
 * from gpsd compiler.h
 */
static inline void memory_barrier(void)
{
  asm volatile ("" : : : "memory");
}

ShmDriver::ShmDriver() :
  Node("shm_driver"),
  shm_unit_(2),
  fixup_date_(false),
  time_ref_topic("time_ref")
{

  // Open SHM, use Deleter to release SHM
  //shm_ = std::unique_ptr<ShmTimeT, std::function<void(ShmTimeT*)>>(attach_shmTime(shm_unit_),
  //    std::bind(ShmDriver::detach_shmTime, this));
}

void ShmDriver::time_ref_cb(const sensor_msgs::msg::TimeReference::SharedPtr msg)
{
  auto lg = this->get_logger();
  const auto &time_ref = msg->time_ref;
  const auto &stamp = msg->header.stamp;

  if (!shm_) {
    RCLCPP_FATAL(lg, "Got time_ref before shm opens.");
    return;
  }

  /* header */
  shm_->mode = 1;
  shm_->nsamples = 3;    // stages of median filter

  shm_->valid = 0;
  shm_->count += 1;
  /* barrier */
  memory_barrier();
  shm_->clockTimeStampSec = time_ref.sec;
  shm_->clockTimeStampUSec = time_ref.nanosec / 1000;
  shm_->clockTimeStampNSec = time_ref.nanosec;
  shm_->receiveTimeStampSec = stamp.sec;
  shm_->receiveTimeStampUSec = stamp.nanosec / 1000;
  shm_->receiveTimeStampNSec = stamp.nanosec;
  shm_->leap = 0;        // LEAP_NOWARNING
  shm_->precision = -1;  // initially 0.5 sec
  memory_barrier();
  /* barrier again */
  shm_->count += 1;
  shm_->valid = 1;

#if 0
  RCLCPP_DEBUG_THROTTLE(lg, RCUTILS_STEADY_TIME, 10, "Got time_ref: %s: %lu.%09lu",
      time_ref->source.c_str(),
      (long unsigned) time_ref->time_ref.sec,
      (long unsigned) time_ref->time_ref.nanosec);
#endif

  /* It is a hack for rtc-less system like Raspberry Pi
   * We check that system time is unset (less than some magic value)
   * and set time.
   *
   * Sudo configuration required for that feature
   * date -d @1234567890: Sat Feb 14 02:31:30 MSK 2009
   */
  if (fixup_date_ && /*ros::Time::now().sec*/0 < 1234567890ULL) {
    rclcpp::Time time_ref_(time_ref);
    set_system_time(time_ref_.seconds());
  }
}

void ShmDriver::set_system_time(const double seconds)
{
  auto lg= this->get_logger();

  RCLCPP_INFO(lg, "Setting system date to: %f", seconds);

  // construct commad: sudo -n date -u -s @1234567890.000
  Poco::Pipe outp, errp;
  Poco::Process::Args args;

  args.push_back("-n");
  args.push_back("date");
  args.push_back("-u");
  args.push_back("-s");
  args.push_back(Poco::format("@%f", seconds));

  Poco::ProcessHandle ph = Poco::Process::launch("sudo", args, 0, &outp, &errp);

  int rc = ph.wait();
  Poco::PipeInputStream outs(outp), errs(errp);
  std::string out, err;

  Poco::StreamCopier::copyToString(outs, out, 4096);
  Poco::StreamCopier::copyToString(errs, err, 4096);

  if (rc == 0) {
    RCLCPP_INFO(lg, "The system date is set.");
    RCLCPP_DEBUG_STREAM(lg, "OUT: " << out);
    RCLCPP_DEBUG_STREAM(lg, "ERR: " << err);
  }
  else {
    RCLCPP_ERROR(lg, "Setting system date failed.");
    RCLCPP_ERROR_STREAM(lg, "OUT: " << out);
    RCLCPP_ERROR_STREAM(lg, "ERR: " << err);
  }
}

/** Map SHM page
 *
 * NOTE: this function did not create SHM like gpsd/ntpshm.c
 */
ShmTimeT* ShmDriver::attach_shmTime(int unit)
{
  auto lg = this->get_logger();

  /* we definitly not root, no IPC_CREAT */
  const int get_flags = 0666;
  key_t key = NTPD_SHM_BASE + unit;

  int shmid = shmget(key, sizeof(shmTime), get_flags);
  if (shmid < 0) {
    RCLCPP_FATAL(lg, "SHM(%d) shmget(0x%08lx, %zd, %o) fail: %s",
        unit, key, sizeof(shmTime), get_flags, strerror(errno));
    return nullptr;
  }

  /* TODO: check number of attached progs */

  void *p = shmat(shmid, 0, 0);
  if (p == (void *) -1) {
    RCLCPP_FATAL(lg, "SHM(%d) shmat(%d, 0, 0) fail: %s", unit, shmid, strerror(errno));
    return nullptr;
  }

  RCLCPP_INFO(lg, "SHM(%d) key 0x%08lx, successfully opened", unit, key);
  return static_cast<ShmTimeT*>(p);
}

void ShmDriver::detach_shmTime(ShmTimeT* shm)
{
  auto lg = this->get_logger();

  if (shm == nullptr)
    return;

  if (shmdt((const void*) shm) == -1) {
    RCLCPP_FATAL(lg, "SHM shmdt(%p) fail: %s", shm, strerror(errno));
  }
}

CLASS_LOADER_REGISTER_CLASS(ShmDriver, rclcpp::Node)


.ycm_extra_conf.py
# This file is NOT licensed under the GPLv3, which is the license for the rest
# of YouCompleteMe.
#
# Here's the license text for this file:
#
# This is free and unencumbered software released into the public domain.
#
# Anyone is free to copy, modify, publish, use, compile, sell, or
# distribute this software, either in source code form or as a compiled
# binary, for any purpose, commercial or non-commercial, and by any
# means.
#
# In jurisdictions that recognize copyright laws, the author or authors
# of this software dedicate any and all copyright interest in the
# software to the public domain. We make this dedication for the benefit
# of the public at large and to the detriment of our heirs and
# successors. We intend this dedication to be an overt act of
# relinquishment in perpetuity of all present and future rights to this
# software under copyright law.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.
#
# For more information, please refer to <http://unlicense.org/>

import os
import ycm_core

# These are the compilation flags that will be used in case there's no
# compilation database set (by default, one is not set).
# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
flags = [
'-Wall',
'-Wextra',
'-Werror',
'-fexceptions',
# THIS IS IMPORTANT! Without a "-std=<something>" flag, clang won't know which
# language to use when compiling headers. So it will guess. Badly. So C++
# headers will be compiled as C headers. You don't want that so ALWAYS specify
# a "-std=<something>".
# For a C project, you would set this to something like 'c99' instead of
# 'c++11'.
'-std=c++14',
# ...and the same thing goes for the magic -x option which specifies the
# language that the files to be compiled are written in. This is mostly
# relevant for c++ headers.
# For a C project, you would set this to 'c' instead of 'c++'.
'-x',
'c++',
'-isystem', "/opt/ros/%s/include" % os.environ.get('ROS_DISTRO', 'indigo'),
'-isystem', '/usr/include/Poco',
'-isystem', '/usr/include',
'-I', '.',
'-I', './include',
'-I', '../devel/include',
# ROS flags
'-DROSCONSOLE_BACKEND_LOG4CXX',
#'-DROS_PACKAGE_NAME="mavros"',
]

# Set this to the absolute path to the folder (NOT the file!) containing the
# compile_commands.json file to use that instead of 'flags'. See here for
# more details: http://clang.llvm.org/docs/JSONCompilationDatabase.html
#
# You can get CMake to generate this file for you by adding:
#   set( CMAKE_EXPORT_COMPILE_COMMANDS 1 )
# to your CMakeLists.txt file.
#
# Most projects will NOT need to set this to anything; you can just change the
# 'flags' list of compilation flags. Notice that YCM itself uses that approach.
compilation_database_folder = ''

if os.path.exists( compilation_database_folder ):
  database = ycm_core.CompilationDatabase( compilation_database_folder )
else:
  database = None

SOURCE_EXTENSIONS = [ '.cpp', '.cxx', '.cc', '.c', '.m', '.mm' ]

def DirectoryOfThisScript():
  return os.path.dirname( os.path.abspath( __file__ ) )


def MakeRelativePathsInFlagsAbsolute( flags, working_directory ):
  if not working_directory:
    return list( flags )
  new_flags = []
  make_next_absolute = False
  path_flags = [ '-isystem', '-I', '-iquote', '--sysroot=' ]
  for flag in flags:
    new_flag = flag

    if make_next_absolute:
      make_next_absolute = False
      if not flag.startswith( '/' ):
        new_flag = os.path.join( working_directory, flag )

    for path_flag in path_flags:
      if flag == path_flag:
        make_next_absolute = True
        break

      if flag.startswith( path_flag ):
        path = flag[ len( path_flag ): ]
        new_flag = path_flag + os.path.join( working_directory, path )
        break

    if new_flag:
      new_flags.append( new_flag )
  return new_flags


def IsHeaderFile( filename ):
  extension = os.path.splitext( filename )[ 1 ]
  return extension in [ '.h', '.hxx', '.hpp', '.hh' ]


def GetCompilationInfoForFile( filename ):
  # The compilation_commands.json file generated by CMake does not have entries
  # for header files. So we do our best by asking the db for flags for a
  # corresponding source file, if any. If one exists, the flags for that file
  # should be good enough.
  if IsHeaderFile( filename ):
    basename = os.path.splitext( filename )[ 0 ]
    for extension in SOURCE_EXTENSIONS:
      replacement_file = basename + extension
      if os.path.exists( replacement_file ):
        compilation_info = database.GetCompilationInfoForFile(
          replacement_file )
        if compilation_info.compiler_flags_:
          return compilation_info
    return None
  return database.GetCompilationInfoForFile( filename )


def FlagsForFile( filename, **kwargs ):
  if database:
    # Bear in mind that compilation_info.compiler_flags_ does NOT return a
    # python list, but a "list-like" StringVec object
    compilation_info = GetCompilationInfoForFile( filename )
    if not compilation_info:
      return None

    final_flags = MakeRelativePathsInFlagsAbsolute(
      compilation_info.compiler_flags_,
      compilation_info.compiler_working_dir_ )
  else:
    relative_to = DirectoryOfThisScript()
    final_flags = MakeRelativePathsInFlagsAbsolute( flags, relative_to )

  return {
    'flags': final_flags,
    'do_cache': True
  }


src/shm_driver_node.cpp
//
// Copyright (c) 2014-2021 Vladimir Ermakov. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
//

/** ROS ntpd driver (via shm)
 *
 * This node will subscribe to any TimeReference message
 * and sends it's data to ntpd.
 *
 * @author Vladimir Ermakov <vooon341@gmail.com>
 * @license BSD
 *
 * Copyright 2014 Vladimir Ermakov.
 * Based on ntpshm.c from gpsd.
 *
 * @file
 */

#include <memory>
#include "ntpd_driver/NtpdShmDriver.hpp"

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ntpd_driver::NtpdShmDriver>());
  rclcpp::shutdown();
  return 0;
}


include/ntpd_driver/ShmDriver.hpp
/** ROS 2 Composable node for ShmDriver
 *
 * This node will subscribe to any TimeReference message
 * and sends it's data to ntpd.
 *
 * @author Vladimir Ermakov <vooon341@gmail.com>
 * @license BSD
 *
 * Copyright 2019 Vladimir Ermakov.
 * Based on ntpshm.c from gpsd.
 *
 * @file
 */

#pragma once

#include <chrono>
#include <memory>
#include <functional>
#include <cstring>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/time_reference.hpp>


/** the definition of shmTime is from ntpd source ntpd/refclock_shm.c */
struct shmTime
{
  int mode; /* 0 - if valid set
             *       use values,
             *       clear valid
             * 1 - if valid set
             *       if count before and after read of values is equal,
             *         use values
             *       clear valid
             */
  volatile int count;
  time_t clockTimeStampSec;
  int clockTimeStampUSec;
  time_t receiveTimeStampSec;
  int receiveTimeStampUSec;
  int leap;
  int precision;
  int nsamples;
  volatile int valid;
  unsigned        clockTimeStampNSec;     /* Unsigned ns timestamps */
  unsigned        receiveTimeStampNSec;   /* Unsigned ns timestamps */
  int             dummy[8];
};

const long int NTPD_SHM_BASE = 0x4e545030;

using ShmTimeT = volatile struct shmTime;


class ShmDriver : public rclcpp::Node
{
  public:
    ShmDriver();

  private:
    rclcpp::Subscription<sensor_msgs::msg::TimeReference>::SharedPtr time_ref_sub_;

    int shm_unit_;
    bool fixup_date_;
    std::string time_ref_topic;

    std::unique_ptr<ShmTimeT, std::function<void(ShmTimeT*)>> shm_;

    void time_ref_cb(const sensor_msgs::msg::TimeReference::SharedPtr msg);
    void set_system_time(const double seconds);

    ShmTimeT* attach_shmTime(int unit);
    void detach_shmTime(ShmTimeT* shm);
};


src/NtpdShmDriver.cpp
//
// Copyright (c) 2014-2021 Vladimir Ermakov. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
//    * Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//    * Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in the
//      documentation and/or other materials provided with the distribution.
//
//    * Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.
//

#include <sys/ipc.h>
#include <sys/shm.h>

// there may be another library, but Poco already used by class_loader,
// so it definetly exist in your system.
#include <Poco/Format.h>
#include <Poco/Pipe.h>
#include <Poco/PipeStream.h>
#include <Poco/Process.h>
#include <Poco/StreamCopier.h>

#include <cstring>
#include <memory>
#include <string>
#include "ntpd_driver/NtpdShmDriver.hpp"


using namespace ntpd_driver; // NOLINT

/**
 * Memory barrier. Unfortunately we can't use C stdatomic.h
 * So only one option: asm magic
 *
 * from gpsd compiler.h
 */
static inline void memory_barrier(void)
{
  asm volatile (""
  :
  :
  : "memory");
}

NtpdShmDriver::NtpdShmDriver(const rclcpp::NodeOptions & options)
: Node("shm_driver", options),
  shm_unit_("shm_unit", 2),
  fixup_date_("fixup_date", false)
{
  this->declare_parameter("shm_unit", shm_unit_.get_parameter_value());
  this->declare_parameter("fixup_date", fixup_date_.get_parameter_value());

  this->get_parameter("shm_unit", shm_unit_);
  this->get_parameter("fixup_date", fixup_date_);

  // Open SHM, use Deleter to release SHM
  shm_ = std::unique_ptr<ShmTimeT, std::function<void(ShmTimeT *)>>(
    attach_shmTime(shm_unit_.as_int()),
    std::bind(&NtpdShmDriver::detach_shmTime, this, std::placeholders::_1));

  time_ref_sub_ = this->create_subscription<sensor_msgs::msg::TimeReference>(
    "time_ref", rclcpp::SensorDataQoS(),
    std::bind(&NtpdShmDriver::time_ref_cb, this, std::placeholders::_1));
}

void NtpdShmDriver::time_ref_cb(const sensor_msgs::msg::TimeReference::SharedPtr msg)
{
  auto lg = this->get_logger();
  auto clock = this->get_clock();

  const auto & time_ref = msg->time_ref;
  const auto & stamp = msg->header.stamp;

  if (!shm_) {
    RCLCPP_FATAL(lg, "Got time_ref before shm opens.");
    return;
  }

  /* header */
  shm_->mode = 1;
  shm_->nsamples = 3;    // stages of median filter

  shm_->valid = 0;
  shm_->count += 1;
  /* barrier */
  memory_barrier();
  shm_->clockTimeStampSec = time_ref.sec;
  shm_->clockTimeStampUSec = time_ref.nanosec / 1000;
  shm_->clockTimeStampNSec = time_ref.nanosec;
  shm_->receiveTimeStampSec = stamp.sec;
  shm_->receiveTimeStampUSec = stamp.nanosec / 1000;
  shm_->receiveTimeStampNSec = stamp.nanosec;
  shm_->leap = 0;          // LEAP_NOWARNING
  shm_->precision = -1;    // initially 0.5 sec
  memory_barrier();
  /* barrier again */
  shm_->count += 1;
  shm_->valid = 1;

  RCLCPP_DEBUG(
    lg, "Got time_ref: %s: %lu.%09lu",
    msg->source.c_str(),
    int64_t(time_ref.sec),
    uint32_t(time_ref.nanosec));

  /* It is a hack for rtc-less system like Raspberry Pi
   * We check that system time is unset (less than some magic value)
   * and set time.
   *
   * date -d @1234567890: Sat Feb 14 02:31:30 MSK 2009
   */
  const rclcpp::Time magic_date(1234567890ULL, 0);
  if (fixup_date_.as_bool() && clock->now() < magic_date) {
    rclcpp::Time time_ref_(time_ref);
    set_system_time(time_ref_.seconds());
  }
}

/** Sets system time by calling `sudo date`
 *
 * Sudo configuration required for that feature
 */
void NtpdShmDriver::set_system_time(const double seconds)
{
  auto lg = this->get_logger();

  RCLCPP_INFO(lg, "Setting system date to: %f", seconds);

  // construct commad: sudo -n date -u -s @1234567890.000
  Poco::Pipe outp, errp;
  Poco::Process::Args args;

  args.push_back("-n");
  args.push_back("date");
  args.push_back("-u");
  args.push_back("-s");
  args.push_back(Poco::format("@%f", seconds));

  Poco::ProcessHandle ph = Poco::Process::launch("sudo", args, 0, &outp, &errp);

  int rc = ph.wait();
  Poco::PipeInputStream outs(outp), errs(errp);
  std::string out, err;

  Poco::StreamCopier::copyToString(outs, out, 4096);
  Poco::StreamCopier::copyToString(errs, err, 4096);

  if (rc == 0) {
    RCLCPP_INFO(lg, "The system date is set.");
    RCLCPP_DEBUG_STREAM(lg, "OUT: " << out);
    RCLCPP_DEBUG_STREAM(lg, "ERR: " << err);
  } else {
    RCLCPP_ERROR(lg, "Setting system date failed.");
    RCLCPP_ERROR_STREAM(lg, "OUT: " << out);
    RCLCPP_ERROR_STREAM(lg, "ERR: " << err);
  }
}

/** Map SHM page
 *
 * NOTE: this function does not create SHM like gpsd/ntpshm.c
 */
ShmTimeT * NtpdShmDriver::attach_shmTime(int unit)
{
  auto lg = this->get_logger();

  /* we definitly not root, no IPC_CREAT */
  const int get_flags = 0666;
  key_t key = NTPD_SHM_BASE + unit;

  int shmid = shmget(key, sizeof(shmTime), get_flags);
  if (shmid < 0) {
    RCLCPP_FATAL(
      lg, "SHM(%d) shmget(0x%08lx, %zd, %o) fail: %s",
      unit, key, sizeof(shmTime), get_flags, strerror(errno));
    return nullptr;
  }

  /* TODO: check number of attached progs */

  const void * shmat_fail_ptr = (void *)-1;  // NOLINT

  void * p = shmat(shmid, 0, 0);
  if (p == shmat_fail_ptr) {
    RCLCPP_FATAL(lg, "SHM(%d) shmat(%d, 0, 0) fail: %s", unit, shmid, strerror(errno));
    return nullptr;
  }

  RCLCPP_INFO(lg, "SHM(%d) key 0x%08lx, successfully opened", unit, key);
  return static_cast<ShmTimeT *>(p);
}

/** Release SHM page
 *
 */
void NtpdShmDriver::detach_shmTime(ShmTimeT * shm)
{
  auto lg = this->get_logger();

  if (shm == nullptr) {
    return;
  }

  if (shmdt((const void *)shm) == -1) {
    RCLCPP_FATAL(lg, "SHM shmdt(%p) fail: %s", shm, strerror(errno));
  }
}


#include <rclcpp_components/register_node_macro.hpp>  // NOLINT
RCLCPP_COMPONENTS_REGISTER_NODE(ntpd_driver::NtpdShmDriver)
